File: prism-sdk/credentials/shared/src/main/scala/io/iohk/atala/prism/credentials/json/JsonBasedCredential.scala
package io.iohk.atala.prism.credentials.json

import java.util.Base64

import io.circe.parser
import io.circe.syntax._
import io.iohk.atala.prism.crypto.{ECTrait, ECSignature, ECPrivateKey}
import io.iohk.atala.prism.util.ArrayOps._
import io.iohk.atala.prism.credentials.Credential
import io.iohk.atala.prism.credentials.errors.CredentialParsingError
import io.iohk.atala.prism.credentials.content.CredentialContent

import io.iohk.atala.prism.credentials.json.implicits._

case class JsonBasedCredential(
    contentBytes: IndexedSeq[Byte],
    content: CredentialContent,
    signature: Option[ECSignature]
) extends Credential {

  override def sign(privateKey: ECPrivateKey)(implicit ec: ECTrait): JsonBasedCredential = {
    copy(
      signature = Some(ec.sign(contentBytes.toByteArray, privateKey))
    )
  }

  override def canonicalForm: String = {
    signature match {
      case Some(signature) =>
        s"${JsonBasedCredential.base64Encoder.encode(contentBytes.toByteArray).asString}" +
          s"${JsonBasedCredential.SEPARATOR}" +
          s"${JsonBasedCredential.base64Encoder.encode(signature.data).asString}"
      case None => contentBytes.toByteArray.asString
    }
  }

}

object JsonBasedCredential {

  val SEPARATOR = '.'
  val base64Decoder = Base64.getUrlDecoder
  val base64Encoder = Base64.getUrlEncoder

  def fromCredentialContent(credentialContent: CredentialContent): JsonBasedCredential = {
    JsonBasedCredential(
      contentBytes = credentialContent.asJson.noSpaces.getBytes.toIndexedSeq,
      content = credentialContent,
      signature = None
    )
  }

  def fromString(credential: String): Either[CredentialParsingError, JsonBasedCredential] = {
    parser.decode[CredentialContent](credential) match {
      case Left(_) =>
        credential.split(SEPARATOR).toList match {
          case content :: signature :: Nil =>
            parser.decode[CredentialContent](base64Decoder.decode(content).asString) match {
              case Left(error) =>
                Left(CredentialParsingError(s"Failed to parse signed credential content: ${error.getMessage}"))
              case Right(credentialContent) =>
                Right(
                  JsonBasedCredential(
                    contentBytes = base64Decoder.decode(content).toIndexedSeq,
                    content = credentialContent,
                    signature = Some(ECSignature(base64Decoder.decode(signature)))
                  )
                )
            }

          case _ =>
            Left(
              CredentialParsingError(
                "Failed to parse signed credential. Expected format: [encoded credential].[encoded signature]"
              )
            )
        }

      case Right(credentialContent) =>
        Right(
          JsonBasedCredential(
            contentBytes = credential.getBytes.toIndexedSeq,
            content = credentialContent,
            signature = None
          )
        )
    }
  }

}

File: prism-kotlin-sdk/credentials/src/commonMain/kotlin/io/iohk/atala/prism/kotlin/credentials/json/JsonBasedCredential.kt
package io.iohk.atala.prism.kotlin.credentials.json

import io.iohk.atala.prism.kotlin.credentials.CredentialParsingError
import io.iohk.atala.prism.kotlin.credentials.PrismCredential
import io.iohk.atala.prism.kotlin.credentials.content.CredentialContent
import io.iohk.atala.prism.kotlin.crypto.EC
import io.iohk.atala.prism.kotlin.crypto.keys.ECPrivateKey
import io.iohk.atala.prism.kotlin.crypto.signature.ECSignature
import io.iohk.atala.prism.kotlin.protos.util.Base64Utils
import kotlin.js.JsExport

@JsExport
data class JsonBasedCredential constructor(
    override val content: CredentialContent,
    override val signature: ECSignature? = null
) : PrismCredential() {

    override val contentBytes = content.fields.toString().encodeToByteArray()

    override val canonicalForm: String =
        if (signature != null) {
            Base64Utils.encode(contentBytes) +
                SEPARATOR +
                Base64Utils.encode(signature.getEncoded())
        } else {
            contentBytes.decodeToString()
        }

    override fun sign(privateKey: ECPrivateKey): PrismCredential {
        return copy(signature = EC.sign(contentBytes, privateKey))
    }

    companion object {
        val SEPARATOR = '.'

        private fun parseUnsignedCredential(credential: String): JsonBasedCredential {
            val credentialContent = CredentialContent.fromString(credential)
            return JsonBasedCredential(credentialContent)
        }

        private fun parseSignedCredential(credential: String): JsonBasedCredential {
            val contentWithSignature = credential.split(SEPARATOR)
            if (contentWithSignature.size == 2) {
                val content = contentWithSignature[0]
                val signature = contentWithSignature[1]

                try {
                    val credentialContent =
                        CredentialContent.fromString(Base64Utils.decode(content).decodeToString())
                    return JsonBasedCredential(
                        content = credentialContent,
                        signature = ECSignature(Base64Utils.decode(signature))
                    )
                } catch (e: Exception) {
                    throw CredentialParsingError("Failed to parse signed credential content: ${e.message}")
                }
            } else {
                throw CredentialParsingError(
                    "Failed to parse signed credential. " +
                        "Expected format: [encoded credential]$SEPARATOR[encoded signature]"
                )
            }
        }

        fun fromString(credential: String): JsonBasedCredential {
            try {
                return parseUnsignedCredential(credential)
            } catch (e: Exception) {
                return parseSignedCredential(credential)
            }
        }
    }
}
